bazel_8: init at 8.1.1

boltzmannrain opened the pull request:
Started with copying `bazel_7` and adapted it, notable changes:
- `bazelDeps` now builds on Darwin in `sandbox=true` which was previously failing on power management tweaks
- using `lndir -silent` for less build logs
- had to disable `module_deps` feature to avoid getting `.. does not depend on a module exporting ..` errors on clang/macos importing intrinsics
- on macos adding `-fno-aligned-allocation` on sdk <10.13 to avoid build error saying it isn't supported on <10.13
- placing under `pkgs/by-name` to pass CI, but still calling from `pkgs/top-level/all-packages.nix` as not sure how to pass darwin sdk and other settings in by-name way

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

boltzmannrain committed:
bazel_8: init at 8.1.1

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
@ofborg build bazel_8

boltzmannrain commented:
on x86_64-darwin following looks like a build agent setup issue?
```
       > FATAL: mkdir('/var/tmp/_bazel__nixbld1'): (error: 13): Permission denied
```

msanft commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 362414`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_8</li>
  </ul>
</details>

timothyklim commented:
https://github.com/bazelbuild/bazel/releases/tag/8.0.1 released yesterday

infinisil commented on ci/OWNERS:
```diff
@@ -302,6 +302,7 @@ pkgs/development/python-modules/buildcatrust/ @ajs124 @lukegb @mweinelt
 
 # Bazel
 /pkgs/development/tools/build-managers/bazel @Profpatsch
+/pkgs/by-name/ba/bazel_8 @Profpatsch
```
Let's not burden people with more pings unless they confirm being okay with it. @Profpatsch do you even still want to help maintain Bazel? If not, you should probably be removed from here.

And independent of whether @Profpatsch stays listed here, I'm suggesting you add yourself @boltzmannrain, since you've done so much great Bazel maintenance recently :D

timothyklim commented:
Since PR created, bazel updated to https://github.com/bazelbuild/bazel/releases/tag/8.1.1

seh commented:
> Where does that occur within the current patch?

Oh, perhaps you meant `module_maps`, which occurs within the _pkgs/by-name/ba/bazel_8/package.nix_ file .

seh commented on pkgs/by-name/ba/bazel_8/package.nix:
```diff
@@ -0,0 +1,808 @@
+          -e "/bazel_build /a\  --repository_disable_download \\\\" \
+          -e "/bazel_build /a\  --announce_rc \\\\" \
+          -e "/bazel_build /a\  --nobuild_python_zip \\\\" \
+          -e "/bazel_build /a\  --features=-module_maps --host_features=-module_maps \\\\" \
```
I found that for what's been ailing me trying to build the _Abseil_ library within Bazel on macOS, just including `--host_features=-module_maps` in my _.bazelrc_ finally got it all to work.

Thank you so much, Dmitry!

boltzmannrain commented:
Updated to 8.1.1, removed ci/OWNERS change, applied couple more build workarounds that showed up. Darwin builds have a new unaddressed `-faligned-allocation` (opposite flag is already added, but maybe not in all contexts)

timothyklim commented:
Can it be merged before 8.2.0?

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 362414`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_8</li>
  </ul>
</details>

philiptaron reviewed (commented):
I read through this PR as a committer who 

a. doesn't use Bazel
b. doesn't know Java that well
c. is just looking to unblock what looks like a stuck PR

So please bear that in mind.

I have a couple big pieces of feedback that would have made it much easier to say "merge immediately".

1. Splitting the mega-derivation into chunks, each of which was `callPackage`d from the main one. That makes it possible for me to say "yeah, that checks out" in chunks, rather than trying to keep the whole 600+ Nix derivation in my head as I'm reading it.
2. Use of patches throughout, so that all of the `sed` calls were reviewable through seeing them in context. Yes, this will be bigger in terms of bytes. Trying to figure out what a multi-line `sed` call is going to do is just super hard.
3. Use of the new Darwin SDK and interaction with recent developments there. More broadly, a lot of this derivation is dedicated to changing Bazel behavior on Darwin, and I found myself wondering whether it was overfitting on a stdenv that no longer is recommended.

There are lots of smaller things I commented on as a read through this PR, but those were the big ones.

I don't actually know if any of that is actually **blocking**, because as previously stated, I'm just here trying to unstick a PR without traction.

hzeller commented:
FWIW, I tested 8.1.1 from this PR on two of projects I maintain ( [bant](https://github.com/hzeller/bant) and [verible](https://github.com/chipsalliance/verible)), both of which are compatible with a wide range of bazel versions from 6.5.0 to 8.1.1.

 * One thing I noticed: the `enableNixHacks` patch does not apply cleanly, so this PR currently only builds in the configuration without. So currently using `bazel_8` when building nix packages does not work as nix-package building is enabling that hack. So only compiled my projects locally with a fresh 8.1.1 build ( `cd my_project; /nix/store/nyc8b16lkh434bqbmnb51ipwl392irkq-bazel-8.1.1/bin/bazel build -c opt ...` )
 * Projects that use 8.1.x seem to implicitly upgrade `rules_cc` to some newer version, or start using `rules_cc` to run the compiler.  In any case, that `rules_cc` is not new enough to contain the patch that makes them nix-compatible ( https://github.com/bazelbuild/rules_cc/pull/306 ) which is not in a released version yet.
 
 So I can _only_ use 8.1.1 from this PR on these projects if I force upgrading to a working `rules_cc` version by adding the following to my projects' `MODULE.bazel`
 
```Python
bazel_dep(name = "rules_cc", version = "0.1.1")
git_override(
    module_name = "rules_cc",
    commit = "d74915024017250e46d95e91a3defc34174effe0",  # get commit of rules_cc PR 306
    remote = "https://github.com/bazelbuild/rules_cc",
)
```

With that, both `bant` and `verible` build.

I wonder if in the 8.1.1 somewhere has its MODULE.bazel that  loads some older `rules_cc`, and maybe we can patch it up to contain that override for it to work ?

hzeller commented:
Uh, looking through the bazel 8.1.1 source code to see what `rules_cc` they use ... I only see references to somewhat ancient ones (also not sure with is implicitly used when bazel is run on a project)

```bash
bazel[release-8.1.1]$ rg rules_cc.*version
src/MODULE.tools:bazel_dep(name = "rules_cc", version = "0.0.17")
src/test/py/bazel/bazel_windows_test.py:            'bazel_dep(name = "rules_cc", version = "0.0.12")',
src/test/py/bazel/launcher_test.py:            'bazel_dep(name = "rules_cc", version = "0.0.12")',
src/test/py/bazel/testdata/runfiles_test/MODULE.bazel.mock:bazel_dep(name = "rules_cc", version = "0.0.9")
src/test/py/bazel/bazel_windows_cpp_test.py:            'bazel_dep(name = "rules_cc", version = "0.0.12")',
site/en/external/module.md:bazel_dep(name = "rules_cc", version = "0.0.1")
site/en/external/overview.md:bazel_dep(name = "rules_cc", version = "0.0.1")
MODULE.bazel:bazel_dep(name = "rules_cc", version = "0.0.17")
```

... not sure if my suggestion to upgrade `rules_cc` to something newer than 0.1.1 is feasible if it is so far off.

hzeller commented:
> ... not sure if my suggestion to upgrade `rules_cc` to something newer than 0.1.1 is feasible if it is so far off.

I am convinced now this is not something we should work around in bazel 8. Bazel 8 actually just introduces a minimal `rules_cc` and that is fine. Other projects might require a newer `rules_cc` because they depend on other things that pull in a rules_cc that has not the fix yet. So I think nothing to block this PR, it is up to the projects compiling things with bazel 8+ to fix that.

So LGTM to me.

boltzmannrain commented:
Marking as draft for now, will probably create a new one after refactoring and making tests work

IanMcShane commented:
Super excited about this PR. Thanks for the effort!

boltzmannrain commented:
Bazel 8.2.1 is much easier to bootstrap-build, so hope to open a new improved PR soon

boltzmannrain commented:
Draft for 8.2.1 https://github.com/NixOS/nixpkgs/pull/400941 I will be refactoring and cleaning it up a fair bit but big-picture is settled for now

boltzmannrain commented:
8.2.1 PR supersedes 8.1.1
